### DISCUSSION TITLE
Putting Material FX, VirtualFX, MFXResources together to release an app

### DISCUSSION THREAD

**Discussion by stefanofornari** [palexdev/MaterialFX#307] [General]
Hi Alex,
I am trying to release a vesion 1.0.0 (or 0.0.1) of my app that uses fundamentally MaterialFX and VirtualFX. I therefore need to get read of the -SNAPSHOTs... I have update the dependency on VirtualedFX to v11.9.0(*)  but it broke the code with MFXResources:

import io.github.palexdev.mfxresources.font.FontResources; -§> not recognized anymore

I currently use these dependencies:

```
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>virtualizedfx</artifactId>
    <version>11.8.3-SNAPSHOT</version>
</dependency>
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>materialfx</artifactId>
    <version>11.13.9</version>
</dependency>
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>mfxcore</artifactId>
    <version>11.1.5</version>
</dependency>
```

Which dependencies should I use without SNAPSHOT to have a compatible app?

Many thanks in advance.

Ste

(*) Please make sure javafx dependencies are transient otherwise it brings in a different version of JavaFX than what one uses in his/her project.

**Comment by palexdev** [comment]
Yes, I'm currently working on a private hobby project, and I noticed this.
In short: I've made a fu....g mess haha

I just updated and released all my components. I completely ditched the old resources, I'm sorry this will probably create problems to you if you are using "mfx-..." for the icons, but it's for the best, it's about time to fix resources conflicts
I've also updated VirtualizedFX to the latest version. MFXCore changed its module name to "mfx.core" and was causing compile issues saying it could not find it

Try with the latest version, 11.16.0

P.S: if you need to find the new icons, you can also go to [FontAwesome Search](https://fontawesome.com/search?o=r&m=free&s=solid)

**Comment by palexdev** [comment]
@stefanofornari this insane configuration finally works for me
![image](https://user-images.githubusercontent.com/16880178/230740287-02a70b34-3002-41b2-89a7-3fcb11855d27.png)

**Comment by stefanofornari** [comment]
You wish :) 
With the below dependencies, I still have the following errors (see https://github.com/stefanofornari/EasyWallet/blob/main/src/main/java/ste/w3/easywallet/ui/SortingIcons.java):

1. package io.github.palexdev.materialfx.font does not exist
2. package io.github.palexdev.mfxresources.font does not exist

If this is expected, can you please give me an hint on how to replace "font" ?

```
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>virtualizedfx</artifactId>
    <version>11.9.1</version>
    <exclusions>
        <exclusion>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-base</artifactId>
        </exclusion>
        <exclusion>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-controls</artifactId>
        </exclusion>
        <exclusion>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-swing</artifactId>
        </exclusion>
        <exclusion>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-fxml</artifactId>
        </exclusion>
        <exclusion>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-media</artifactId>
        </exclusion>
        <exclusion>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-web</artifactId>
        </exclusion>
        <exclusion>
            <groupId>org.openjfx</groupId>
            <artifactId>javafx-graphics</artifactId>
        </exclusion>
    </exclusions>
</dependency>
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>materialfx</artifactId>
    <version>11.16.0</version>
</dependency>
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>mfxcore</artifactId>
    <version>11.3.0</version>
</dependency>
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>mfxresources</artifactId>
    <version>11.3.2</version>
</dependency>
<dependency>
    <groupId>io.github.palexdev</groupId>
    <artifactId>mfxcomponents</artifactId>
    <version>11.16.2</version>
</dependency>
```

**Reply by palexdev** [reply]
The font package doesn't exist anymore, it is now "fonts"
Just replace where needed


Edit: here's the new class
```java
package ste.w3.easywallet.ui; 
  
  
import io.github.palexdev.mfxresources.fonts.IconDescriptor;
import io.github.palexdev.mfxresources.fonts.MFXFontIcon;
import io.github.palexdev.mfxresources.fonts.fontawesome.FontAwesomeSolid;

 /** 
  * 
  */ 
 public enum SortingIcons { 
     ASCENDING(FontAwesomeSolid.CARET_UP), 
     DESCENDING(FontAwesomeSolid.CARET_DOWN), 
     NONE(FontAwesomeSolid.MINUS); 
  
     public final IconDescriptor symbol; 
  
     SortingIcons(IconDescriptor symbol) { 
         this.symbol = symbol; 
     } 
  
     public MFXFontIcon newIcon() { 
         return new MFXFontIcon(symbol); 
     } 
 }
```


Also you can bump the dependencies, there are new minor changes:
- mfxresources 11.3.3
- mfxcomponents 11.6.2
- materialfx 11.16.1
- virtualizedfx 11.9.2

Edit 2: It should not be necessary to exclude JavaFX when importing VirtualizedFX, the dependencies are already transitive

**Comment by stefanofornari** [comment]
Thanks! One step forward. I bumped the version numbers and used the code above and it now compiles. I have the following exception though:

```
java.lang.IllegalArgumentException: Cannot register this When construct as the given observable is already being observed
	io.github.palexdev.mfxcore.observables.When.register(When.java:108)
	io.github.palexdev.mfxcore.observables.OnChanged.listen(OnChanged.java:192)
	io.github.palexdev.virtualizedfx.table.defaults.SimpleTableCell.lambda$init$1(SimpleTableCell.java:125)
	com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:181)
	com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	javafx.beans.property.ReadOnlyObjectPropertyBase.fireValueChangedEvent(ReadOnlyObjectPropertyBase.java:80)
	javafx.beans.property.ReadOnlyObjectWrapper.fireValueChangedEvent(ReadOnlyObjectWrapper.java:102)
	javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:113)
	javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
	io.github.palexdev.virtualizedfx.table.defaults.SimpleTableCell.setColumn(SimpleTableCell.java:189)
	io.github.palexdev.virtualizedfx.table.defaults.SimpleTableCell.updateColumn(SimpleTableCell.java:151)
	io.github.palexdev.virtualizedfx.table.defaults.DefaultTableRow.lambda$takeFromCacheOrCreate$4(DefaultTableRow.java:216)
	java.base/java.util.Optional.orElseGet(Optional.java:364)
	io.github.palexdev.virtualizedfx.table.defaults.DefaultTableRow.takeFromCacheOrCreate(DefaultTableRow.java:229)
	io.github.palexdev.virtualizedfx.table.defaults.DefaultTableRow.init(DefaultTableRow.java:59)
	io.github.palexdev.virtualizedfx.table.TableState.addRow(TableState.java:516)
	io.github.palexdev.virtualizedfx.table.TableManager.init(TableManager.java:127)
	io.github.palexdev.virtualizedfx.table.TableManager.onChange(TableManager.java:256)
	io.github.palexdev.virtualizedfx.table.VirtualTableSkin.onItemsChanged(VirtualTableSkin.java:215)
	com.sun.javafx.collections.ListListenerHelper$SingleChange.fireValueChangedEvent(ListListenerHelper.java:164)
	com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73)
	javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:239)
	javafx.collections.ListChangeBuilder.commit(ListChangeBuilder.java:482)
	javafx.collections.ListChangeBuilder.endChange(ListChangeBuilder.java:541)
	javafx.collections.ObservableListBase.endChange(ObservableListBase.java:211)
	javafx.collections.ModifiableObservableListBase.addAll(ModifiableObservableListBase.java:109)
	ste.w3.easywallet.ui.LedgerController.lambda$initialize$4(LedgerController.java:113)
```

running this code (which was working before):

```
Platform.runLater(() -> {
    table.getItems().clear();
    table.getItems().addAll(source.page);
    table.scrollToRow((int)(source.page.size()-source.pageSize()));
});
```

**Reply by palexdev** [reply]
That's probably a bug since When now doesn't allow registering multiple times on the same property, I must have missed it since it didn't show in the tests
I'll have to fix it and release a new version

**Reply by stefanofornari** [reply]
mmhh... I thing there is really a mess with the versions... I used mfxcore 11.3 as per your screenshot above, but I can not find that version in the repository; have you pushed the changes already?
I reverted to v 11.1.5 and I've got it back working.

**Reply by palexdev** [reply]
I think this is an issue with Maven, if you can try adding it as a local jar

**Reply by stefanofornari** [reply]
What I meant is that if I look at the report, I do not see any sign of version 11.3; nor there are tags. Just version 11.1.5..

**Reply by palexdev** [reply]
Oh I see maybe you are looking on the wrong old repository still
The new MFXCore is here on the MaterialFX repository in the `rewrite` branch

**Reply by stefanofornari** [reply]
oh, I see. Definitely I was looking at the old repository...

**Comment by palexdev** [comment]
To be honest, I'd like them to be capable of registering multiple times...but the disposal would be more complicated
I'll try to find a solution

**Comment by stefanofornari** [comment]
Huston, we have a problem :)

I have compared the app with the new versions (left) with the old version (right). There are a few L&F differences, are they wanted so we need to do things differently?
What is more concerning though, are the performance. The new versions are way slower than the old ones :(


https://user-images.githubusercontent.com/43138/230780892-52fc6c62-8ff7-4e63-9b44-386ef0ab9e7a.mp4

**Reply by palexdev** [reply]
No there's definitely something wrong going on, but I need a reproducer to investigate and fix it

**Reply by stefanofornari** [reply]
I'll try to write a mini project to reproduce the issues.

**Reply by palexdev** [reply]
Thanks, it could be anything
A theming issue, conflicts, layout issues... anything
I need an app I can debug to check where lag is happening exactly

**Comment by stefanofornari** [comment]
Here you go: https://github.com/stefanofornari/mfxfixme

The main branch contains the pom with the old versions
The newmfx branch contains the pom with the new versions

If you run EditDialogDemo, you can notice:

1. the layout is screwed
2. performance issue (I hope, I see it on my PC which is not super fast... let me know)

Let me know if you see the same and if I can be of any help.

Ste

PS: sorry I use maven, but you should easily be able to adapt it to gradle

**Reply by palexdev** [reply]
I swear I will never understand how you can't run Gradle projects and I can't run Maven ones...like wtf haha
I'm trying to figure out how to run it :(

Edit: got it, added a Launcher class ^^'

**Reply by palexdev** [reply]
@stefanofornari 
I have some news:
1) On the new branch I see the button is not styled, it's because you didn't add the theme on the Scene not a big deal
2) I see the performance lag when pressing the button, BUT... this is because you are using them the wrong way. You see, JavaFX lacks APIs (and I'm talking native APIs) to build dialogs, in fact even their dialog implementation is just an extension of Stage. Building a Stage, though, is a very expensive operation and one should absolutely avoid building them on the fly.
If you have a View/Controller/wathever, and you know you'll need to show a dialog, it's recommended to build an instance of the dialog (in this case MFXStageDialog) and keep it as an object reference. When you need to show different things, just change the Stage/Dialog content (in this case AbstractMFXDialog). Even better, you could also keep instances of the dialog content and just make sure to init/reset them when you need. In short, the fewer objects you create with `new`  when you need to show a dialog, the better the performance
3) I didn't notice any other style inconsistencies, in your video I see scroll bars, but you haven't provided an example for them in the demo you gave me, I could test it if you can tell me what are they, MFXScrollPane? VirtualScrollPane?
4) I removed all the `exclusion` directives from the VirtualizedFX section in the pom, and I saw no difference in the dependencies, like I said previously, VirtualizedFX already declares all its dependencies as transitive, there's no need for that

These are the changes/comments/suggestions I added:
![Screenshot from 2023-04-11 13-21-32](https://user-images.githubusercontent.com/16880178/231157531-4d4ee7e7-3504-40ae-b435-79d6261a1ce5.png)

**Reply by stefanofornari** [reply]
Thanks @palexdev! however, you are working around the problem. Ok, I am doing it in the wrong way, but at least it shows a noticeable performance issue in the new version. What's the cause of it? 
I will double check on the button later, but do you see the dialog showing up correctly? Here I see the correct layout with the old versions, but a screwed layout with the new ones. This should be a different issue than the performance.

**Reply by stefanofornari** [reply]
I mean this:

OLD
![image](https://user-images.githubusercontent.com/43138/231354604-76268b1d-9057-4f98-bbb4-d6fdfed5e744.png)

NEW:
![image](https://user-images.githubusercontent.com/43138/231354948-a01c9d2e-29e5-4c7e-8393-7c3f1ecbeae6.png)

**Reply by palexdev** [reply]
> Thanks @palexdev! however, you are working around the problem. Ok, I am doing it in the wrong way, but at least it shows a noticeable performance issue in the new version. What's the cause of it? I will double check on the button later, but do you see the dialog showing up correctly? Here I see the correct layout with the old versions, but a screwed layout with the new ones. This should be a different issue than the performance.

If I have to guess, it may be due to the fact that dialogs, since are separate Scenes, add the whole Theme stylesheet on themselves.
You can fix this by clearing the stylesheets list, and just add the ones you need (depending on the controls inside the dialog)
Ultimately, the API should be improved.

**Reply by stefanofornari** [reply]
I am not sure I understand... you do not have to guess, you can debug the app. You are doing a ton of things and your work is exceptional, but I wonder if it is ready for a real app. I think my app is quite simple in terms of UI. It's a good use case for MFX and I am happy to help if I can. But I need also to move forward with the app. What's your suggestion then? shall I fork the old versions, find a combination that works and stick to that? Or shall we use my app as a test bed for MFX and therefore fix the issues and not finding workaround? I think the latter is better in both interested, but it requires your buy-in as well.

**Reply by palexdev** [reply]
> I am not sure I understand... you do not have to guess, you can debug the app. You are doing a ton of things and your work is exceptional, but I wonder if it is ready for a real app. I think my app is quite simple in terms of UI. It's a good use case for MFX and I am happy to help if I can. But I need also to move forward with the app. What's your suggestion then? shall I fork the old versions, find a combination that works and stick to that? Or shall we use my app as a test bed for MFX and therefore fix the issues and not finding workaround? I think the latter is better in both interested, but it requires your buy-in as well.

In fact, I don't have to guess. Compared to old releases, the only thing I changed that also affected dialogs was the theming. So, as I already said, the performance issue is probably due to how dialogs are integrated with the theming API. Upon their creation (exactly here [L48](https://github.com/palexdev/MaterialFX/blob/b3106326b71b580590a2d38e30bdffafaab4a425/materialfx/src/main/java/io/github/palexdev/materialfx/dialogs/AbstractMFXDialog.java#L48)) the entire theme is applied on them. This is good because the user doesn't have to worry about adding the styles he needs, and it's bad because it's heavy on performance, since now JavaFX has to parse the entire stylesheet.

Since I can't test it properly because on my PCs I don't see any difference, probably because they are powerful enough that it doesn't matter, I was suggesting you trying to remove the themes, like this:
```java
// Just an example, adapt to your needs
MFXStageDialog sd ...;
sd.getContent().getStylesheets().setAll(...); // Set the ones you need based on the controls inside the dialog
```

Also, I'm not saying I won't fix it, (I said, and I quote `Ultimately, the API should be improved.`), For now whatever workaround makes your app work as you need is worth implementing.

**Reply by stefanofornari** [reply]
You like to make me confused don't you?! :) following the discussion flow, I was commenting on the stylesheets issue, not performance; you then said "I have to guess".... anyway, it does not really matter :) maybe we should create separate issues, let me know and I can do it.

On performance: you mentioned multiple times how much important performance are for the UI, IMHO you should take into proper consideration that the current Theming implementation has a big impact. Although you may not notice it on a your machine in this specific case, the overhead is there and can show up in other circumstances. You have the opportunity to use my boomer machine as a proxy if you need more information. Just let me know how I can help.

The most concerning thing however is the styles screw-up. I tried your suggestion, but it does not bring any particular improvement:
![image](https://user-images.githubusercontent.com/43138/231872364-f89ad439-c12e-4453-b1cb-8c672960af57.png)
if now the way to style MFX app is different than before, I am ok to follow the new way, but I need to know/understand how to do it. If there is already some documentation feel free to point me to that.

PS: for now I manage to create a release with the old versions, getting rid of the use of font icons, but it is a pity because I can not test the new version.

**Comment by palexdev** [comment]
Just to confirm my hypotheses. Here's the same issue from a project I'm working on: [Video](https://github.com/palexdev/MaterialFX/assets/16880178/7c6ad27f-a9b8-4bba-9fe3-ab90abfc4a6a)

The first run uses `addAll` to add the stylesheets on the dialog and when clicking the button the app freezes for a while and then the dialog is shown.
In the second run, after changing `addAll` to `setAll`, the dialog opens almost instantly. It's clear that adding the entire theme to the dialog ([L48](https://github.com/palexdev/MaterialFX/blob/b3106326b71b580590a2d38e30bdffafaab4a425/materialfx/src/main/java/io/github/palexdev/materialfx/dialogs/AbstractMFXDialog.java#L48)) is a terrible thing that leads to an awful performance.
For now, the workaround is to remove the themes by overriding them with `setAll`. A fix for this... I don't know, I'll have to think about it.
